File: src/main/java/cses/dynamic_programming/RemovingDigits.java
package cses.dynamic_programming;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public class RemovingDigits {
    static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = Integer.parseInt(scanner.nextLine());

        List<Integer> answer = solution(n);
        System.out.println(answer.size());
    }

    public static List<Integer> solution(int n) {
        int curN = n;
        List<Integer> result = new ArrayList<>();

        while (curN > 0) {
            List<Integer> digits = getDigits(10, curN);
            int maxDigit = Collections.max(digits);
            curN -= maxDigit;
            result.add(curN);
        }

        return result;
    }

    public static List<Integer> getDigits(int base, int number) {
        List<Integer> digits = new ArrayList<>();
        int prevPlaceValue = 1;
        int placeValue = base;
        int prev = 0;

        while (true) {
            int mod = number % placeValue;
            int digit = (mod - prev) / prevPlaceValue;
            digits.add(digit);
            if (placeValue > number) break;
            prev = mod;
            prevPlaceValue = placeValue;
            placeValue *= base;
        }

        Collections.reverse(digits);
        return digits;
    }
}

File: src/main/kotlin/cses/dynamic_programming/RemovingDigits.kt
package cses.dynamic_programming

fun main() {
//    val n = readln().toInt()
    val n = 27

    val answer = solution(n)

    println(answer)
}

private fun solution(n: Int): Int {
    var curN = n
    val result = mutableListOf<Int>()

    while (curN > 0) {
        val digits = getDigits(10, curN)

        curN -= digits.max()
        result.add(curN)
    }

    return result.size
}

private fun getDigits(base: Int, number: Int): List<Int> {
    val digits = mutableListOf<Int>()
    var prevPlaceValue = 1
    var placeValue = base
    var prev = 0
    while (true) {
        val mod = number % placeValue
        val digit = (mod - prev) / prevPlaceValue
        digits.add(digit)
        if (placeValue > number) break
        prev = mod
        prevPlaceValue = placeValue
        placeValue *= base
    }
    return digits.reversed()
}
